File: src/oauth/pkce.ts
/**
 * OAuth PKCE (S256) Implementation
 * Requirements: 21.2, 26.4
 *
 * Implements PKCE (Proof Key for Code Exchange) using S256 method.
 * Based on RFC 7636.
 */

import { createHash, randomBytes } from 'crypto';
import { CodeChallengeMethod } from './types.js';

/**
 * Characters allowed in code verifier (unreserved characters per RFC 7636)
 */
const UNRESERVED_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';

/**
 * Default code verifier length
 */
const DEFAULT_VERIFIER_LENGTH = 64;

/**
 * Minimum and maximum verifier length per RFC 7636
 */
const MIN_VERIFIER_LENGTH = 43;
const MAX_VERIFIER_LENGTH = 128;

/**
 * Generate a cryptographically random code verifier
 *
 * @param length - Length of the verifier (default: 64, min: 43, max: 128)
 * @returns A random code verifier string
 */
export function generateCodeVerifier(length: number = DEFAULT_VERIFIER_LENGTH): string {
  // Clamp length to valid range
  const validLength = Math.max(MIN_VERIFIER_LENGTH, Math.min(MAX_VERIFIER_LENGTH, length));

  const bytes = randomBytes(validLength);
  let verifier = '';

  for (let i = 0; i < validLength; i++) {
    verifier += UNRESERVED_CHARS[bytes[i] % UNRESERVED_CHARS.length];
  }

  return verifier;
}

/**
 * Generate a code challenge from a code verifier using S256 method
 *
 * S256: BASE64URL(SHA256(code_verifier))
 *
 * @param codeVerifier - The code verifier to hash
 * @returns Base64URL encoded SHA256 hash of the verifier
 */
export function generateCodeChallenge(codeVerifier: string): string {
  // Calculate SHA256 hash
  const hash = createHash('sha256').update(codeVerifier, 'ascii').digest();

  // Convert to base64url encoding (no padding)
  const base64 = hash.toString('base64');
  const base64url = base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');

  return base64url;
}

/**
 * Verify a code verifier against a code challenge
 *
 * @param codeVerifier - The code verifier from the token request
 * @param codeChallenge - The code challenge from the authorization request
 * @param method - The code challenge method (only 'S256' is supported)
 * @returns True if the verifier matches the challenge
 */
export function verifyCodeChallenge(
  codeVerifier: string,
  codeChallenge: string,
  method: CodeChallengeMethod
): boolean {
  if (method !== 'S256') {
    throw new Error('Only S256 code challenge method is supported');
  }

  // Generate challenge from verifier and compare
  const expectedChallenge = generateCodeChallenge(codeVerifier);
  return expectedChallenge === codeChallenge;
}

/**
 * Validate code verifier format
 *
 * @param codeVerifier - The code verifier to validate
 * @returns True if the verifier has valid format
 */
export function isValidCodeVerifier(codeVerifier: string): boolean {
  if (!codeVerifier) {
    return false;
  }

  // Check length
  if (codeVerifier.length < MIN_VERIFIER_LENGTH || codeVerifier.length > MAX_VERIFIER_LENGTH) {
    return false;
  }

  // Check characters (unreserved characters only)
  const validPattern = /^[A-Za-z0-9\-._~]+$/;
  return validPattern.test(codeVerifier);
}

/**
 * Validate code challenge format
 *
 * @param codeChallenge - The code challenge to validate
 * @returns True if the challenge has valid format
 */
export function isValidCodeChallenge(codeChallenge: string): boolean {
  if (!codeChallenge) {
    return false;
  }

  // S256 challenge should be base64url encoded SHA256 hash (43 characters)
  if (codeChallenge.length !== 43) {
    return false;
  }

  // Check base64url format (no padding)
  const validPattern = /^[A-Za-z0-9\-_]+$/;
  return validPattern.test(codeChallenge);
}

File: rust/crates/sage-integrations/src/pkce.rs
//! PKCE (S256) — port of `src/google-oauth/pkce.ts` (RFC 7636).

use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine as _};
use rand::RngCore;
use sha2::{Digest, Sha256};

const UNRESERVED_CHARS: &[u8] =
    b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
const DEFAULT_VERIFIER_LENGTH: usize = 64;
const MIN_VERIFIER_LENGTH: usize = 43;
const MAX_VERIFIER_LENGTH: usize = 128;

/// Generate a random code verifier (length clamped to [43, 128], default 64).
pub fn generate_code_verifier(length: usize) -> String {
    let valid_length = length.clamp(MIN_VERIFIER_LENGTH, MAX_VERIFIER_LENGTH);
    let mut bytes = vec![0u8; valid_length];
    rand::thread_rng().fill_bytes(&mut bytes);
    bytes
        .iter()
        .map(|b| UNRESERVED_CHARS[(*b as usize) % UNRESERVED_CHARS.len()] as char)
        .collect()
}

pub fn default_verifier() -> String {
    generate_code_verifier(DEFAULT_VERIFIER_LENGTH)
}

/// `BASE64URL(SHA256(code_verifier))`, no padding. Verifier hashed as ASCII bytes.
pub fn generate_code_challenge(code_verifier: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(code_verifier.as_bytes());
    let digest = hasher.finalize();
    // base64 then url-safe substitution (matches the TS replace chain).
    let b64 = STANDARD_NO_PAD.encode(digest);
    b64.replace('+', "-").replace('/', "_")
}

pub fn verify_code_challenge(code_verifier: &str, code_challenge: &str) -> bool {
    generate_code_challenge(code_verifier) == code_challenge
}

pub fn is_valid_code_verifier(code_verifier: &str) -> bool {
    let len = code_verifier.len();
    if !(MIN_VERIFIER_LENGTH..=MAX_VERIFIER_LENGTH).contains(&len) {
        return false;
    }
    code_verifier
        .bytes()
        .all(|b| b.is_ascii_alphanumeric() || matches!(b, b'-' | b'.' | b'_' | b'~'))
}

pub fn is_valid_code_challenge(code_challenge: &str) -> bool {
    if code_challenge.len() != 43 {
        return false;
    }
    code_challenge
        .bytes()
        .all(|b| b.is_ascii_alphanumeric() || matches!(b, b'-' | b'_'))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rfc7636_challenge_vector() {
        // RFC 7636 Appendix B example.
        let verifier = "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk";
        let challenge = generate_code_challenge(verifier);
        assert_eq!(challenge, "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM");
        assert!(verify_code_challenge(verifier, &challenge));
        assert!(is_valid_code_verifier(verifier));
        assert!(is_valid_code_challenge(&challenge));
    }

    #[test]
    fn verifier_length_clamping_and_charset() {
        assert_eq!(generate_code_verifier(10).len(), MIN_VERIFIER_LENGTH);
        assert_eq!(generate_code_verifier(9999).len(), MAX_VERIFIER_LENGTH);
        assert_eq!(default_verifier().len(), DEFAULT_VERIFIER_LENGTH);
        assert!(is_valid_code_verifier(&default_verifier()));
    }

    #[test]
    fn invalid_inputs_rejected() {
        assert!(!is_valid_code_verifier("short"));
        assert!(!is_valid_code_verifier(&"a".repeat(200)));
        assert!(!is_valid_code_verifier(
            "has space aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
        ));
        assert!(!is_valid_code_challenge("tooshort"));
    }
}
